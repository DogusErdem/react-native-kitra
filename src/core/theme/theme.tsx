import { Dispatch, SetStateAction, createContext, useEffect, useMemo, useState } from 'react';
import { useColorScheme } from 'react-native';
import type { FCC, ThemeType } from '../../types';

type ThemeContextType = {
  theme: { [index: string]: string }
  updateTheme: (theme: { dark: { [index: string]: string }, light: { [index: string]: string } }) => void,
  setColorScheme: Dispatch<SetStateAction< 'light' | 'dark' | undefined>>,
  colorScheme?: 'light' | 'dark' | undefined,

}

const ThemeContext = createContext<ThemeContextType>({} as ThemeContextType);

const ThemeProvider: FCC = ({ children }) => {
  const [colorScheme, setColorScheme] = useState<'light' | 'dark' | undefined>('dark');
  const [theme, setTheme] = useState(defaultTheme);
  function updateTheme(newTheme: { dark: { [index: string]: string }, light: { [index: string]: string } }) {
    setTheme(prev => ({
      ...prev,
      dark: {
        ...prev.dark,
        ...newTheme?.dark,

      },
      light: {
        ...prev.light,
        ...newTheme?.light,
      },

    }));
  }

  const GetColors = () => theme[colorScheme!];

  const contextValue = useMemo(() => ({ theme: GetColors(), updateTheme, setColorScheme, colorScheme }), [theme, colorScheme]);

  return (
    <ThemeContext.Provider value={contextValue}>
      {children}
    </ThemeContext.Provider>
  );
};

export { ThemeProvider, ThemeContext };

const text = {
  black: '#000000',
  lightBlack: '#404040',
  grey: '#9A9A9A',
  white: '#FFFFFF',
  darkWhite: '#FDFCFF',
  lightGrey: '#F5F4F6',
};

const status = {
  focused: '#654FA6',
  error: '#FF3434',
  succes: '#09CE63',
  disabledLight: '#D7D1E9',
  disabledDark: '#BDBCBF',
  disabledLightDark: '#F6F6F6',
};

export const defaultTheme: ThemeType = {
  dark: {
    primary: '#8973CD',
    primary5: '#F9F8FD',
    primary15: '#EDEAF8',
    primary30: '#DCD5F0',
    secondary: '#82B98E',
    tertiary: '#67A7C1',
    ...text,
    ...status,
  },
  light: {
    primary: '#8973CD',
    primary5: '#F9F8FD',
    primary15: '#F3F1FA',
    primary30: '#DCD5F0',
    secondary: '#82B98E',
    tertiary: '#67A7C1',
    ...text,
    ...status,
  },
};
