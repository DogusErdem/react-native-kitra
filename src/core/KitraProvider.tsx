import type { ComponentType, ForwardedRef } from 'react';
import React from 'react';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
import Animated from 'react-native-reanimated';
import { StyleProp, ViewStyle } from 'react-native';

import { ThemeProvider } from './theme/theme';
import { TypographyContext, TypographyProvider } from './typography/typography';
import { NotificationContextType, NotificationProvider } from './notification/notification';

const Feather = require('react-native-vector-icons/Feather').default;
const Ionicons = require('react-native-vector-icons/Ionicons').default;
const AntDesign = require('react-native-vector-icons/AntDesign').default;

Animated.addWhitelistedNativeProps({ text: true });
const messageTypes = (theme:any) => ({
  SUCCESS: {
    backgroundColor: theme.colors.status.successLight,
    icon: <Feather name="check" size={25} color={theme.colors.status.success} />,
  },
  INFO: {
    backgroundColor: theme.colors.system.primary15,
    icon: <Ionicons name="information-circle-outline" size={25} color={theme.colors.system.primary} />,
  },
  DANGER: {
    backgroundColor: theme.colors.status.errorLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.error} />,
  },
  WARNING: {
    backgroundColor: theme.colors.status.warningLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.warning} />,
  },
});

type KitraProviderType= {
  children?:any,
  limit?:number,
  notificationRef?:React.Ref<NotificationContextType>,
  messageType?:(theme?:any, colorScheme?:'dark' | 'light') => {[key:string]: {
      backgroundColor:string,
    icon?:React.ReactNode
    onPress?:()=>void
  }},
  notificationCcontainerStyle?:StyleProp<ViewStyle>,
  customView?:({ header, type, message, theme }:
    {header:string, type:string, message:string, theme:any})=>React.ReactNode}

export const KitraProvider =
({ children, messageType = messageTypes, notificationCcontainerStyle, limit, customView,
  notificationRef }: KitraProviderType) => (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <ThemeProvider>
        <TypographyProvider>
          <NotificationProvider
            messageType={theme => (messageType ? messageType(theme) : theme)}
            notificationCcontainerStyle={notificationCcontainerStyle}
            limit={limit}
            customView={item => customView?.(item)}
            ref={notificationRef}
          >
            {children}
          </NotificationProvider>
        </TypographyProvider>
      </ThemeProvider>
    </GestureHandlerRootView>
);

export function applyDefaults<T>(Component: ComponentType<T>) {
  return React.forwardRef((props: T, ref: ForwardedRef<any>) => (
    <TypographyContext.Consumer>
      {typography =>
        <Component ref={ref} typography={typography?.typography} {...props} />
          }
    </TypographyContext.Consumer>
  ));
}
