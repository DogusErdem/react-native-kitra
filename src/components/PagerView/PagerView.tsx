import React, { Dispatch, ReactNode, SetStateAction, useEffect, useRef, useState } from 'react';
import Animated, { FadeIn, runOnJS, SharedValue, useAnimatedStyle, useDerivedValue, useSharedValue } from 'react-native-reanimated';
import { StyleProp, StyleSheet, TextStyle, TouchableOpacity, View, ViewStyle } from 'react-native';
import { default as PagerViewComponent, PagerViewOnPageScrollEvent, PagerViewProps as PWProps } from 'react-native-pager-view';
import type { FCCWD, IconComponentProps, IconType, PagerViewProps } from '../../types';
import { applyDefaults } from '../../core/KitraProvider';
import Divider from '../Divider/Divider';
import type Icon from '../Icons/Icon';

const style = StyleSheet.create({
  container: { flex: 1 },
  headersContainer: { flexDirection: 'row', padding: 5, borderRadius: 10, marginLeft: 'auto', marginRight: 'auto', width: 315, height: 50 },
  headerContainer: { flex: 1, justifyContent: 'center', alignItems: 'center' },
  headerText: { margin: 10, textAlign: 'center', fontSize: 12, fontWeight: '500', flexDirection: 'row' },
  slider: { position: 'absolute', borderRadius: 8, margin: 5, alignItems: 'center' },
  pageContainer: { height: '100%', width: '100%', flex: 1 },
  divider: { position: 'absolute', bottom: 0, height: 4, margin: 5 },
});

export type TabItemProps = {
  item: ReactNode,
  index: number,
  theme:{[index: string]: string} | undefined,
  refPager: React.RefObject<PagerViewComponent>,
  setSize: Dispatch<SetStateAction<{ height: number, width: number }>>,
  slideValue: SharedValue<number>
  headerTextColor: {select:string|undefined, default:string|undefined},
  headerTextStyle?:StyleProp<TextStyle>,
  icons?: React.ReactElement<typeof Icon>[],
  selectPage:number
}
const TabItem:FCCWD<TabItemProps> = ({ theme, icons, selectPage, typography, item, index, refPager, setSize, slideValue, headerTextStyle, headerTextColor }) => {
  const textColorStyle = useAnimatedStyle(() => {
    if (index - slideValue.value < 0.3 && index - slideValue.value > -0.5) { return { color: headerTextColor.select || theme?.white }; }
    return { color: headerTextColor.default || theme?.primary };
  });

  return (
    <TouchableOpacity
      key={index}
      testID={`tab_button_${index}`}
      onLayout={({ nativeEvent }) =>
        setSize({ height: nativeEvent.layout.height, width: nativeEvent.layout.width })}
      style={[style.headerContainer]}
      onPress={() => { refPager.current?.setPage(index); }}
    >
      <View style={{ flexDirection: 'row', alignItems: 'center' }}>
        {/* @ts-ignore */}
        {icons && icons[index] && React.cloneElement(icons[index], { color: selectPage === index ? headerTextColor.select : headerTextColor.default })}
        {/* @ts-ignore */}
        <Animated.Text style={[{ fontSize: typography?.body.medium.fontSize, lineHeight: typography?.body.medium.lineHeight }, style.headerText, textColorStyle, headerTextStyle]}>{item.key}</Animated.Text>
      </View>
    </TouchableOpacity>
  );
};

const AnimatedPager = Animated.createAnimatedComponent(PagerViewComponent);

const PagerView:FCCWD<PagerViewProps&PWProps> =
({ theme,
  typography,
  children,
  containerStyle,
  pageContainerStyle,
  headerTextColor,
  headerTextStyle,
  injectPagerRef = false,
  headerSliderStyle,
  headerContainerStyle,
  showDivider = true,
  dividerStyle,
  dividerColor,
  icons }) => {
  const refPager = useRef<PagerViewComponent>(null);
  const slideValue = useSharedValue(0);
  const [size, setSize] = useState<{ height: number, width: number }>({ height: 0, width: 0 });
  const [selectPage, setSelectPage] = useState(0);
  const headerSliderAnimated = useAnimatedStyle(() => (
    { left: (slideValue.value * (size?.width ? size.width : 0)) ? (slideValue.value * (size?.width ? size.width : 0)) : 0 }
  ));

  const onPageScroll = (e: PagerViewOnPageScrollEvent) => {
    if (e.nativeEvent.offset < 1) { slideValue.value = +e.nativeEvent.offset + e.nativeEvent.position; }
  };

  return (
    <View style={[style.container, containerStyle]}>
      <View style={[{ backgroundColor: theme?.white }, style.headersContainer, headerContainerStyle]}>
        {React.Children.map(children, (item, index) => (
          <TabItem
            theme={theme}
            selectPage={selectPage}
            item={item}
            icons={icons}
            index={index}
            refPager={refPager}
            setSize={setSize}
            slideValue={slideValue}
            typography={typography}
            headerTextStyle={headerTextStyle}
            headerTextColor={headerTextColor || { select: theme?.white, default: theme?.primary }}
          />
        ))
        }
        <Animated.View style={[style.slider, { backgroundColor: theme?.focused, height: size?.height, width: size?.width, zIndex: 100 }, headerSliderAnimated, headerSliderStyle]} />
        {showDivider &&
        <Divider style={[style.divider, dividerStyle]} color={dividerColor} />
 }
      </View>

      <AnimatedPager
        ref={refPager}
        entering={FadeIn}
        onPageScroll={onPageScroll}
        overScrollMode="never"
        overdrag={false}
        style={[{ flex: 1 }, pageContainerStyle]}
        initialPage={0}
        onPageSelected={e => setSelectPage(e.nativeEvent.position)}
      >

        {React.Children.map(children, (item, index) => {
          if (React.isValidElement(item)) {
            return (
              <View style={[style.pageContainer]} key={index + 1}>
                {injectPagerRef ? React.cloneElement(item.props.children, { ref: refPager }) : item}
              </View>
            );
          }
          return null;
        })}
      </AnimatedPager>
    </View>
  );
};
export default applyDefaults(PagerView);
