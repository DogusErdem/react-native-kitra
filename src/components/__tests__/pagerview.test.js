/* eslint-disable no-undef */

import { fireEvent, render, act, screen } from '@testing-library/react-native';
import { Text, View } from 'react-native';
import { KitraProvider } from '../../core/KitraProvider';
import PagerView from '../PagerView/PagerView';
import { defaultTheme } from '../../core/theme/theme';

const renderWithContext = children => render(<KitraProvider>{children}</KitraProvider>);
const onPageScrollMock = jest.fn();
const onPageSelectedMock = jest.fn();

const RenderItem = initialPage => (
  <PagerView
    initialPage={initialPage}
    showDivider
    dividerColor={defaultTheme.dark.grey}
    pageContainerStyle={{ height: 500 }}
    dividerStyle={{ height: 2 }}
    headerContainerStyle={{ backgroundColor: defaultTheme.dark.background, width: '100%' }}
    headerTextColor={{ select: defaultTheme.dark.white, default: defaultTheme.dark.disabledDark }}
    headerSliderStyle={[{ backgroundColor: defaultTheme.dark.white, bottom: 0, height: 2, marginTop: 20 }]}
    containerStyle={{ paddingHorizontal: 24 }}
    testID="pager_view"
  >
    <View key="Screen 1" testID="screen_1">
      <Text>
        Test1
      </Text>
    </View>
    <View key="Screen 2" testID="screen_2">
      <Text>
        Test2
      </Text>
    </View>
  </PagerView>
);
describe('PagerView', () => {
  it('renders PagerView', () => {
    const tree = renderWithContext(<PagerView />);
    expect(tree).toMatchSnapshot();
  });
  it('renders PagerView with not react element', () => {
    const tree = renderWithContext(<PagerView>Test</PagerView>);
    expect(tree).toMatchSnapshot();
  });
  it('tab items button press', async () => {
    renderWithContext(<RenderItem />);
    const screen1 = screen.getByTestId('screen_1');
    const screen2 = screen.getByTestId('screen_2');

    const tabButton = screen.getByTestId('tab_button_0');

    fireEvent.press(tabButton);
    act(() => {
      fireEvent(tabButton, 'layout', {
        nativeEvent: { layout: { height: 40 } },
      });
    });

    expect(screen1).toBeVisible();

    const tabButton1 = screen.getByTestId('tab_button_1');
    fireEvent.press(tabButton1);

    expect(screen2).toBeVisible();
  });
});
