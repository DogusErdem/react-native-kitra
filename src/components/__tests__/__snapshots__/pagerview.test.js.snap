// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PagerView renders PagerView 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "#FFFFFF",
          },
          {
            "borderRadius": 10,
            "flexDirection": "row",
            "height": 50,
            "marginLeft": "auto",
            "marginRight": "auto",
            "padding": 5,
            "width": 315,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "borderRadius": 8,
              "margin": 5,
              "position": "absolute",
            },
            {
              "backgroundColor": "#654FA6",
              "height": 0,
              "width": 0,
              "zIndex": 100,
            },
            {
              "left": 0,
            },
            undefined,
          ]
        }
      />
      <View
        style={
          [
            {
              "alignItems": "center",
              "alignSelf": "center",
              "justifyContent": "center",
            },
            {
              "width": "100%",
            },
            [
              {
                "bottom": 0,
                "height": 4,
                "margin": 5,
                "position": "absolute",
              },
              undefined,
            ],
          ]
        }
      >
        <View
          style={
            {
              "borderColor": "#DCD5F0",
              "borderStyle": "solid",
              "borderWidth": 0.5,
              "width": "100%",
            }
          }
        />
      </View>
    </View>
    <RNCViewPager
      entering={BaseAnimationMock {}}
      initialPage={0}
      layoutDirection="ltr"
      onMoveShouldSetResponderCapture={[Function]}
      onPageScroll={[Function]}
      onPageScrollStateChanged={[Function]}
      onPageSelected={[Function]}
      overScrollMode="never"
      overdrag={false}
      style={
        [
          {
            "flex": 1,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    layout={BaseAnimationMock {}}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 10,
          "height": 100,
          "paddingHorizontal": 24,
          "width": "100%",
          "zIndex": 100,
        },
        {
          "alignSelf": "center",
          "top": 150,
        },
        undefined,
        {
          "backgroundColor": "transparent",
          "marginTop": -110,
          "position": "absolute",
        },
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "opacity": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 100,
            "height": 28,
            "justifyContent": "center",
            "marginLeft": 18,
            "marginTop": 21,
            "width": 28,
            "zIndex": 100,
          }
        }
      />
      <View
        style={
          {
            "flex": 1,
            "zIndex": 100,
          }
        }
      >
        <Text
          ellipsizeMode="middle"
          numberOfLines={3}
          style={
            [
              {
                "fontWeight": "500",
                "marginTop": 17,
                "paddingHorizontal": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        />
        <Text
          ellipsizeMode="middle"
          numberOfLines={3}
          style={
            [
              {
                "fontWeight": "400",
                "marginTop": 6,
                "paddingHorizontal": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`PagerView renders PagerView with not react element 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "#FFFFFF",
          },
          {
            "borderRadius": 10,
            "flexDirection": "row",
            "height": 50,
            "marginLeft": "auto",
            "marginRight": "auto",
            "padding": 5,
            "width": 315,
          },
          undefined,
        ]
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onLayout={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "opacity": 1,
          }
        }
        testID="tab_button_0"
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              [
                {
                  "fontSize": 14,
                  "lineHeight": 21,
                },
                {
                  "flexDirection": "row",
                  "fontSize": 12,
                  "fontWeight": "500",
                  "margin": 10,
                  "textAlign": "center",
                },
                {
                  "color": "#FFFFFF",
                },
                undefined,
              ]
            }
          />
        </View>
      </View>
      <View
        style={
          [
            {
              "alignItems": "center",
              "borderRadius": 8,
              "margin": 5,
              "position": "absolute",
            },
            {
              "backgroundColor": "#654FA6",
              "height": 0,
              "width": 0,
              "zIndex": 100,
            },
            {
              "left": 0,
            },
            undefined,
          ]
        }
      />
      <View
        style={
          [
            {
              "alignItems": "center",
              "alignSelf": "center",
              "justifyContent": "center",
            },
            {
              "width": "100%",
            },
            [
              {
                "bottom": 0,
                "height": 4,
                "margin": 5,
                "position": "absolute",
              },
              undefined,
            ],
          ]
        }
      >
        <View
          style={
            {
              "borderColor": "#DCD5F0",
              "borderStyle": "solid",
              "borderWidth": 0.5,
              "width": "100%",
            }
          }
        />
      </View>
    </View>
    <RNCViewPager
      entering={BaseAnimationMock {}}
      initialPage={0}
      layoutDirection="ltr"
      onMoveShouldSetResponderCapture={[Function]}
      onPageScroll={[Function]}
      onPageScrollStateChanged={[Function]}
      onPageSelected={[Function]}
      overScrollMode="never"
      overdrag={false}
      style={
        [
          {
            "flex": 1,
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    layout={BaseAnimationMock {}}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 10,
          "height": 100,
          "paddingHorizontal": 24,
          "width": "100%",
          "zIndex": 100,
        },
        {
          "alignSelf": "center",
          "top": 150,
        },
        undefined,
        {
          "backgroundColor": "transparent",
          "marginTop": -110,
          "position": "absolute",
        },
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "opacity": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 100,
            "height": 28,
            "justifyContent": "center",
            "marginLeft": 18,
            "marginTop": 21,
            "width": 28,
            "zIndex": 100,
          }
        }
      />
      <View
        style={
          {
            "flex": 1,
            "zIndex": 100,
          }
        }
      >
        <Text
          ellipsizeMode="middle"
          numberOfLines={3}
          style={
            [
              {
                "fontWeight": "500",
                "marginTop": 17,
                "paddingHorizontal": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        />
        <Text
          ellipsizeMode="middle"
          numberOfLines={3}
          style={
            [
              {
                "fontWeight": "400",
                "marginTop": 6,
                "paddingHorizontal": 20,
              },
              {
                "color": "#FFFFFF",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
